SuspendingClock on Windows does not suspend

grynspan opened the pull request:
<!-- What's in this pull request? -->
This PR replaces the implementation of `SuspendingClock` on Windows with one that uses `QueryUnbiasedInterruptTimePrecise()`, which _does_ suspend.

I also updated `swift_get_clock_res()` to produce accurate values on Windows instead of a hard-coded 1µs resolution.

Raymond Chen gives a little bit of info about the various timer APIs [here](https://devblogs.microsoft.com/oldnewthing/20170921-00/?p=97057).

<!--
If this pull request resolves any GitHub issues, link them.
For information about linking a pull request to an issue, see:
https://docs.github.com/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->
Resolves #63224.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

grynspan committed:
SuspendingClock on Windows does not suspend #63224

grynspan commented:
@swift-ci please test

compnerd reviewed (approved):
Thanks for fixing this!  I think that the commit message could be better (the issue helped me understand the change).

grynspan committed:
Fix inaccurate values for swift_get_clock_res() on Windows

grynspan commented on stdlib/public/Concurrency/Clock.cpp:
```diff
@@ -132,8 +124,10 @@ switch (clock_id) {
+      QueryPerformanceFrequency(&freq);
       *seconds = 0;
-      *nanoseconds = 1000;
+      *nanoseconds = 1000000000 / freq.QuadPart;
```
`QueryPerformanceFrequency()` never reports zero post-Windows XP. If we're concerned about that, we can check.

grynspan commented on stdlib/public/Concurrency/Clock.cpp:
```diff
@@ -159,7 +153,7 @@ switch (clock_id) {
 #elif defined(_WIN32)
       *seconds = 0;
-      *nanoseconds = 1000;
+      *nanoseconds = 100;
```
`QueryUnbiasedInterruptTimePrecise()` has a unit of 100ns, so its highest possible resolution is 100ns. Neat.

grynspan commented:
@swift-ci please test

grynspan committed:
QueryUnbiasedInterruptTimePrecise() isn't always available so it must be dynamically looked up. Also, fix the math after calling it, and avoid floating-point math on the continuous
clock.

grynspan commented:
@swift-ci please test

al45tair commented on stdlib/public/Concurrency/Clock.cpp:
```diff
@@ -52,14 +54,14 @@ void swift_get_time(
+      // count by 1,000,000,000 to get nanosecond resolution. By multiplying
+      // first, we maintain high precision. The resulting value is the tick
+      // count in nanoseconds.
+      long long ns = (count.QuadPart * 1000000000) / freq.QuadPart;
```
Couldn't `count.QuadPart * 1000000000` overflow? I think the previous code might have been the way it was in order to avoid that problem?

grynspan replied:
`count.QuadPart` is a 64-bit value and the frequency is CPU-bound (clock can't go brrr if CPU doesn't go brrr.) The counter is a 64-bit value and, unless my math is way off, would only overflow if the system had years of uptime.

However, that's a bit handwavey—we could approach this problem in a couple of different ways:

1. Instead of using `QueryPerformanceCounter()`, use `QueryInterruptTimePrecise()` and get a value in units of 100ns, which will overflow after about 58,455 years, but which is slower to call than `QueryPerformanceCounter()`.
2. Use 128-bit multiplication and division to get the nanosecond count. Windows has various bits of 128-bit math API we could use here. I don't know what sort of performance overhead this API has, but it's arithmetic so I'd hope not _too_ bad.

grynspan replied:
I used 128-bit math.

al45tair commented on stdlib/public/Concurrency/Clock.cpp:
```diff
@@ -87,18 +89,33 @@ void swift_get_time(
+      // QueryUnbiasedInterruptTimePrecise() was added in Windows 10 and is, as
+      // the name suggests, more precise than QueryUnbiasedInterruptTime().
+      // However, despite being declared in Windows' headers, we must look it up
+      // dynamically at runtime.
```
I assume this is because we don't have Windows 10 as a minimum target? (That is, `_WIN32_WINNT` would need to be defined to something else?)

Perhaps the comment should say _why_ it needs to be dynamically looked-up.

grynspan replied:
Exactly. I'll edit the comment to make the connection to Windows 10 clearer.

grynspan replied:
I tried at desk, and from what I can see, even on Windows 10 this symbol is not exported normally. I don't need to change `_WIN32_WINNT`, the symbol is _declared_, but it's not available at link time. It is available when dynamically looking it up using `GetProcAddress()`, and it's fully documented on MSDN. I'll note the issue in the comment.

al45tair replied:
If it's not available at link time, it might just be that we don't link against the relevant `.lib` file?

grynspan replied:
Ah, I'm not crazy! swift-corelibs-dispatch has the same problem. See [this](https://github.com/apple/swift-corelibs-libdispatch/pull/498) PR from that repo.

al45tair reviewed (commented):
Not sure about the change to the division; could you check that, please.

grynspan committed:
- Add more comments explaining why we use GetProcAddress().
- Use 128-bit math with QueryPerformanceCounter() to avoid overflowing.
- Use swift::fatalError() instead of abort() for bad clock IDs.
- Use digit separators (C++14 feature, makes large integers easier to read.)

grynspan commented:
@swift-ci please test

grynspan commented on stdlib/public/Concurrency/Clock.cpp:
```diff
@@ -52,14 +56,16 @@ void swift_get_time(
+      // count in nanoseconds. Use 128-bit math to avoid overflowing.
+      DWORD64 hi = 0;
+      DWORD64 lo = _umul128(count.QuadPart, 1'000'000'000, &hi);
+      DWORD64 ns = _udiv128(hi, lo, freq.QuadPart, nullptr);
```
`_udiv128()` generates a GPF if the result can't fit in a 64-bit integer, but if that happens that means the computer's been up for more than 2^64 nanoseconds… that's 584 years. If anybody runs into that problem, please let my great-great-great-great-great-great-great-great-great-great-great-great-great-great-great-great-great-great-great-great-great-great-great-great-great-great-grandkid know.

grynspan committed:
Do 128-bit math with clang extensions instead of MSVC++ extensions/intrinsics

grynspan commented:
@swift-ci please test windows

grynspan committed:
Use swift_Concurrency_fatalError() instead of swift::fatalError() (oops)

grynspan commented:
@swift-ci please test windows

grynspan committed:
Remove commented-out MSVC++ 128-bit math

grynspan commented:
@swift-ci please test

grynspan committed:
Don't need immintrin.h

grynspan commented:
@swift-ci please test

grynspan commented:
@swift-ci please test linux

al45tair commented on stdlib/public/Concurrency/Clock.cpp:
```diff
@@ -52,14 +56,15 @@ void swift_get_time(
+      auto quadPart = static_cast<unsigned __int128>(count.QuadPart);
+      auto ns = (quadPart * 1'000'000'000) / freq.QuadPart;
+      *seconds = ns / 1'000'000'000;
+      *nanoseconds = ns % 1'000'000'000;
```
Nice :-)

grynspan replied:
Yeah, the "which compiler builds this file?" part tripped me up!

al45tair reviewed (approved):
LGTM

grynspan commented on stdlib/public/Concurrency/Clock.cpp:
```diff
@@ -132,8 +154,10 @@ switch (clock_id) {
+      QueryPerformanceFrequency(&freq);
       *seconds = 0;
-      *nanoseconds = 1000;
+      *nanoseconds = 1'000'000'000 / freq.QuadPart;
```
That would imply the tick duration was less than one nanosecond. It's conceivable that future hardware could support that, but the finest resolution `swift_get_clock_res()` can represent is one nanosecond, so we'd need to refactor this entire function anyway, presumably to support returning attoseconds.)

grynspan committed:
Get rid of redundant timespecs

grynspan commented:
@swift-ci please test

phausler reviewed (approved):
This looks great to me, the cleanup for non-windows platforms looks correct (and definitely cleaner). 

Thanks for taking this on!

grynspan commented:
> This looks great to me, the cleanup for non-windows platforms looks correct (and definitely cleaner).
> 
> Thanks for taking this on!

Always happy to help! (**Danger:** I might noodle on this code a bit more in the future if I see ways to improve it further.)

grynspan committed:
Every platform has time.h, so remove the HAS_TIME check

grynspan commented:
@swift-ci please test

grynspan commented:
@swift-ci please test windows

grynspan commented:
@swift-ci please test windows

grynspan commented:
The reported Windows failure is not related to this PR.

grynspan commented:
@swift-ci please test windows
